File: frontend-v2/src/services/biddingGame.ts
import api from './api'

export interface AgentConfig {
  name: string
  strategy: string
  tech_score: number
  cost_base: number
  profit_target: number
  risk_preference: number
  discount_belief_mean: number
  discount_belief_std: number
}

export interface BiddingScenarioConfig {
  budget: number
  scoring_method: string
  tech_weight: number
  price_weight: number
  k_value: number
  sensitivity: number
  tax_rate: number
}

export interface SimulationConfig {
  n_simulations: number
  method: string
  iterative_rounds?: number
  learning_rate?: number
  exploration_rate?: number
  convergence_threshold?: number
}

export interface AllianceConfig {
  leader: string
  supporters: string[]
  coordination_type: string
  leader_bonus: number
  discount_spread: number
}

export interface CoalitionConfig {
  alliances: AllianceConfig[]
  enabled: boolean
  profit_redistribution: boolean
}

export interface CoalitionAgentEffect {
  agent_name: string
  role: string
  alliance_id: number
  discount_shift: number
  effective_discount: number
}

export interface CoalitionResult {
  alliances: string[]
  agent_effects: CoalitionAgentEffect[]
  alliance_count: number
  coordination_type_breakdown: Record<string, number>
}

export interface BiddingGameSimulatePayload {
  project_id?: string
  scenario: BiddingScenarioConfig
  our_agent: AgentConfig
  competitor_agents: AgentConfig[]
  simulation_config?: SimulationConfig
  coalition_config?: CoalitionConfig
}

export interface BiddingGameHistoryLearningPayload {
  project_id?: string
  competitor_names: string[]
  limit?: number
}

export interface CompetitorHistoryProfile {
  name: string
  discount_belief_mean: number
  discount_belief_std: number
  sample_count: number
  source_breakdown: Record<string, number>
}

export interface BiddingGameHistoryLearningResponse {
  project_id?: string
  profiles: CompetitorHistoryProfile[]
  total_records_scanned: number
  matched_competitor_count: number
  message: string
}

export interface OptimalBidResult {
  recommended_price: number
  recommended_discount: number
  win_probability: number
  expected_profit: number
  confidence_interval: [number, number]
}

export interface SimulationStats {
  n_simulations: number
  win_rate_at_optimal: number
  avg_profit_at_optimal: number
  median_rank: number
  p10_rank: number
  p90_rank: number
}

export interface SensitivityResult {
  most_sensitive_param: string
  price_elasticity: number
}

export interface NashEquilibriumResult {
  found: boolean
  our_optimal_discount: number
  equilibrium_type: string
}

export interface BayesianBeliefUpdate {
  agent_name: string
  prior_mean: number
  prior_std: number
  posterior_mean: number
  posterior_std: number
  observed_samples: number[]
  n_observations: number
  belief_shift: number
}

export interface DiscountDataPoint {
  discount: number
  win_rate: number
}

export interface ProfitDataPoint {
  discount: number
  avg_profit: number
}

export interface BayesianDetail {
  agent_name: string
  prior_mean: number
  prior_std: number
  posterior_mean: number
  posterior_std: number
  belief_shift: number
  n_observations: number
}

export interface RawSimulationData {
  discount_winrate_curve?: DiscountDataPoint[]
  discount_profit_curve?: ProfitDataPoint[]
  bayesian_details?: BayesianDetail[]
  iterative_rounds?: IterativeRoundSnapshot[]
  strategy_evolutions?: StrategyEvolutionSnapshot[]
}

export interface IterativeRoundResult {
  round_no: number
  our_discount: number
  our_price_score: number
  our_total_score: number
  our_rank: number
  won: boolean
  competitor_discounts: number[]
  competitor_scores: number[]
  profit: number
}

export interface AgentStrategyEvolution {
  agent_name: string
  initial_discount_mean: number
  final_discount_mean: number
  strategy_shift: number
  rounds_played: number
  wins: number
  learning_curve: number[]
}

export interface IterativeGameResult {
  rounds: IterativeRoundResult[]
  strategy_evolutions: AgentStrategyEvolution[]
  convergence_round: number
  final_optimal_discount: number
  final_win_probability: number
  final_expected_profit: number
  insights: string[]
}

export interface IterativeRoundSnapshot {
  round_no: number
  our_discount: number
  our_rank: number
  profit: number
  won: boolean
}

export interface StrategyEvolutionSnapshot {
  agent_name: string
  initial_discount_mean: number
  final_discount_mean: number
  strategy_shift: number
  learning_curve: number[]
}

export interface BiddingGameSimulateResponse {
  optimal_bid: OptimalBidResult
  simulation_stats: SimulationStats
  sensitivity: SensitivityResult
  nash_equilibrium: NashEquilibriumResult
  bayesian_updates: BayesianBeliefUpdate[]
  coalition_result?: CoalitionResult | null
  game_insights: string[]
  raw_simulation_data?: RawSimulationData
  iterative_result?: IterativeGameResult | null
}

export async function simulateBiddingGame(payload: BiddingGameSimulatePayload): Promise<BiddingGameSimulateResponse> {
  const res = await api.post<BiddingGameSimulateResponse>('/bidding-game/simulate', payload)
  return res.data
}

export async function learnBiddingGameHistory(payload: BiddingGameHistoryLearningPayload): Promise<BiddingGameHistoryLearningResponse> {
  const res = await api.post<BiddingGameHistoryLearningResponse>('/bidding-game/history-learning', payload)
  return res.data
}

export interface ABTestStrategyGroup {
  label: string
  our_agent: AgentConfig
}

export interface ABTestPayload {
  project_id?: string
  scenario: BiddingScenarioConfig
  strategy_groups: ABTestStrategyGroup[]
  competitor_agents: AgentConfig[]
  n_simulations: number
  coalition_config?: CoalitionConfig
}

export interface ABTestStrategyResult {
  label: string
  optimal_discount: number
  recommended_price: number
  win_probability: number
  expected_profit: number
  median_rank: number
  avg_profit: number
  confidence_interval: [number, number]
  win_rate_curve: { discount: number; win_rate: number }[]
  profit_curve: { discount: number; avg_profit: number }[]
}

export interface ABTestComparison {
  best_strategy: string
  best_win_probability: number
  best_profit: number
  win_rate_ranking: { label: string; win_probability: number }[]
  profit_ranking: { label: string; expected_profit: number }[]
  significance_tests: {
    comparison: string
    z_score: number
    p_value: number
    significant_at_005: boolean
    win_rate_diff: number
  }[]
  recommendation: string
}

export interface ABTestResponse {
  strategy_results: ABTestStrategyResult[]
  comparison: ABTestComparison
  insights: string[]
}

export async function runABTest(payload: ABTestPayload): Promise<ABTestResponse> {
  const res = await api.post<ABTestResponse>('/bidding-game/ab-test', payload)
  return res.data
}

File: backend-v2/app/schemas/bidding_game.py
from pydantic import BaseModel, Field
from typing import Optional


class AgentConfig(BaseModel):
    name: str
    strategy: str = Field(default="balanced", description="aggressive / conservative / balanced / accomplice")
    tech_score: float = Field(ge=0, le=100, default=75)
    cost_base: float = Field(ge=0, default=0)
    profit_target: float = Field(ge=0, default=15)
    risk_preference: float = Field(ge=-1, le=1, default=0.0)
    discount_belief_mean: float = Field(ge=0, le=1, default=0.5)
    discount_belief_std: float = Field(ge=0, le=0.5, default=0.1)


class BiddingScenarioConfig(BaseModel):
    budget: float = Field(ge=0, default=0)
    scoring_method: str = Field(default="linear", description="linear / vertexRandomK / vertexFixedK")
    tech_weight: float = Field(ge=0, le=1, default=0.5)
    price_weight: float = Field(ge=0, le=1, default=0.5)
    k_value: float = Field(ge=0, le=100, default=95)
    sensitivity: float = Field(ge=0, default=2)
    tax_rate: float = Field(ge=0, le=1, default=0.06)


class SimulationConfig(BaseModel):
    n_simulations: int = Field(ge=100, le=10000, default=1000)
    method: str = Field(
        default="monte_carlo",
        description="monte_carlo / nash / sensitivity / bayesian / iterative",
    )
    iterative_rounds: int = Field(ge=5, le=100, default=12)
    learning_rate: float = Field(ge=0.01, le=1, default=0.18)
    exploration_rate: float = Field(ge=0, le=0.5, default=0.12)
    convergence_threshold: float = Field(ge=0.001, le=0.1, default=0.01)


class BayesianBeliefUpdate(BaseModel):
    agent_name: str
    prior_mean: float
    prior_std: float
    posterior_mean: float
    posterior_std: float
    observed_samples: list[float] = []
    n_observations: int = 0
    belief_shift: float = 0.0


class AllianceConfig(BaseModel):
    leader: str
    supporters: list[str] = Field(min_length=1)
    coordination_type: str = Field(
        default="high_bid_escort",
        description="high_bid_escort / price_padding / bracket",
    )
    leader_bonus: float = Field(ge=0, le=1, default=0.0)
    discount_spread: float = Field(ge=0.01, le=0.5, default=0.06)


class CoalitionConfig(BaseModel):
    alliances: list[AllianceConfig] = Field(default_factory=list)
    enabled: bool = False
    profit_redistribution: bool = False


class CoalitionAgentEffect(BaseModel):
    agent_name: str
    role: str = ""
    alliance_id: int = 0
    discount_shift: float = 0.0
    effective_discount: float = 0.0


class CoalitionResult(BaseModel):
    alliances: list[str] = []
    agent_effects: list[CoalitionAgentEffect] = []
    alliance_count: int = 0
    coordination_type_breakdown: dict[str, int] = Field(default_factory=dict)


class BiddingGameSimulateRequest(BaseModel):
    project_id: Optional[str] = None
    scenario: BiddingScenarioConfig
    our_agent: AgentConfig
    competitor_agents: list[AgentConfig] = []
    simulation_config: SimulationConfig = SimulationConfig()
    coalition_config: Optional[CoalitionConfig] = None


class BiddingGameHistoryLearningRequest(BaseModel):
    project_id: Optional[str] = None
    competitor_names: list[str] = Field(default_factory=list)
    limit: int = Field(ge=1, le=200, default=50)


class CompetitorHistoryProfile(BaseModel):
    name: str
    discount_belief_mean: float = Field(ge=0, le=1)
    discount_belief_std: float = Field(ge=0, le=0.5)
    sample_count: int = 0
    source_breakdown: dict[str, int] = Field(default_factory=dict)


class BiddingGameHistoryLearningResponse(BaseModel):
    project_id: Optional[str] = None
    profiles: list[CompetitorHistoryProfile] = Field(default_factory=list)
    total_records_scanned: int = 0
    matched_competitor_count: int = 0
    message: str = ""


class OptimalBidResult(BaseModel):
    recommended_price: float
    recommended_discount: float
    win_probability: float
    expected_profit: float
    confidence_interval: list[float] = Field(min_length=2, max_length=2)


class SimulationStats(BaseModel):
    n_simulations: int
    win_rate_at_optimal: float
    avg_profit_at_optimal: float
    median_rank: float
    p10_rank: float
    p90_rank: float


class SensitivityResult(BaseModel):
    most_sensitive_param: str = ""
    price_elasticity: float = 0.0


class NashEquilibriumResult(BaseModel):
    found: bool = False
    our_optimal_discount: float = 0.0
    equilibrium_type: str = ""


class IterationRoundResult(BaseModel):
    round_no: int
    our_discount: float
    our_price_score: float
    our_total_score: float
    our_rank: int
    won: bool
    competitor_discounts: list[float] = []
    competitor_scores: list[float] = []
    profit: float = 0.0


class AgentStrategyEvolution(BaseModel):
    agent_name: str
    initial_discount_mean: float
    final_discount_mean: float
    strategy_shift: float = 0.0
    rounds_played: int = 0
    wins: int = 0
    learning_curve: list[float] = []


class IterativeGameResult(BaseModel):
    rounds: list[IterationRoundResult] = []
    strategy_evolutions: list[AgentStrategyEvolution] = []
    convergence_round: int = 0
    final_optimal_discount: float = 0.0
    final_win_probability: float = 0.0
    final_expected_profit: float = 0.0
    insights: list[str] = []


class BiddingGameSimulateResponse(BaseModel):
    optimal_bid: OptimalBidResult
    simulation_stats: SimulationStats
    sensitivity: SensitivityResult
    nash_equilibrium: NashEquilibriumResult
    bayesian_updates: list[BayesianBeliefUpdate] = []
    coalition_result: Optional[CoalitionResult] = None
    game_insights: list[str] = []
    raw_simulation_data: Optional[dict] = None
    iterative_result: Optional[IterativeGameResult] = None


class ABTestStrategyGroup(BaseModel):
    label: str = Field(default="A", description="Strategy group label, e.g. A / B / C")
    our_agent: AgentConfig


class ABTestRequest(BaseModel):
    project_id: Optional[str] = None
    scenario: BiddingScenarioConfig
    strategy_groups: list[ABTestStrategyGroup] = Field(min_length=2, max_length=5)
    competitor_agents: list[AgentConfig] = []
    n_simulations: int = Field(ge=100, le=10000, default=500)
    coalition_config: Optional[CoalitionConfig] = None


class ABTestStrategyResult(BaseModel):
    label: str
    optimal_discount: float
    recommended_price: float
    win_probability: float
    expected_profit: float
    median_rank: float
    avg_profit: float
    confidence_interval: list[float] = Field(min_length=2, max_length=2)
    win_rate_curve: list[dict] = Field(default_factory=list)
    profit_curve: list[dict] = Field(default_factory=list)


class ABTestComparison(BaseModel):
    best_strategy: str = ""
    best_win_probability: float = 0.0
    best_profit: float = 0.0
    win_rate_ranking: list[dict] = Field(default_factory=list)
    profit_ranking: list[dict] = Field(default_factory=list)
    significance_tests: list[dict] = Field(default_factory=list)
    recommendation: str = ""


class ABTestResponse(BaseModel):
    strategy_results: list[ABTestStrategyResult] = []
    comparison: ABTestComparison = ABTestComparison()
    insights: list[str] = []
